translation of StrictMode (#569)
reactjs/fr.react.dev: src/content/reference/react/StrictMode.md

---

## Référence {/*reference*/}

### `<StrictMode>` {/*strictmode*/}

> Utilisez `StrictMode` pour activer des comportements de développement et des avertissements supplémentaires pour l'arbre des composants à l'intérieur :

Reviewer: ```suggestion
Utilisez `StrictMode` pour activer des comportements de développement et des avertissements supplémentaires pour l'arbre des composants que vous placez à l'intérieur :
```

);
```

> [Voir d'autres exemples plus bas.](#usage)

Reviewer: ```suggestion
[Voir d'autres exemples plus bas](#usage).
```

Author: Oui mais alors non ?

Jusqu'à présent, on avait dit (dans mes précédentes PR), que si le lien contient toute la phrase, alors on y intègre la ponctuation, ce qui est ici le cas.

La règle a changé ?

Reviewer: Ah non ça c'est pour la graisse et l'italique.

Si tu as trace de la même règle pour la ponctuation je veux bien voir car c'est une erreur.

100% des liens "autres exemples plus bas" sont formatés comme ça, notamment.

Author: Ah, en regardant les anciennes PR, je ne retrouve pas ce que je pensais y trouver, donc laissons tomber 😅

Les comportements suivants sont activés en développement par le mode strict :

> - Vos composants seront [rendus une fois de plus](#fixing-bugs-found-by-double-rendering-in-development) afin de trouver les bugs causés par des rendus impurs.

Reviewer: ```suggestion
- Vos composants feront [un rendu supplémentaire](#fixing-bugs-found-by-double-rendering-in-development) afin de trouver les bugs causés par des rendus impurs.
```

> - Vos composants [réexécuteront les Effets une fois de plus](#fixing-bugs-found-by-re-running-effects-in-development) afin de détecter les bugs causés par l'absence de nettoyage d'Effet.

Reviewer: ```suggestion
- Vos composants [exécuteront les Effets une fois supplémentaire](#fixing-bugs-found-by-re-running-effects-in-development) afin de détecter les bugs causés par l'absence de nettoyage d'Effet.
```

re + "une fois de plus" = 2x ? 3x ? J'harmonise avec l'élément d'avant.

- Vos composants [seront contrôlés pour l'utilisation d'API dépréciées](#fixing-deprecation-warnings-enabled-by-strict-mode).

#### Props {/*props*/}

> `StrictMode` n'accepte aucune props.

Reviewer: ```suggestion
`StrictMode` ne prend aucune prop.
```

);
```

Nous recommandons d'enrober toute votre appli dans le mode strict, en particulier pour les nouvelles applis. Si vous utilisez un framework qui appelle [`createRoot`](/reference/react-dom/client/createRoot) à votre place, consultez sa documentation pour savoir comment activer le mode strict.

> 

Reviewer: ```suggestion
```

Bien que les vérifications en mode strict **ne s'exécutent que durant le développement**, elles vous aident à trouver des bugs qui existent déjà dans votre code et qui peuvent être difficiles à reproduire de façon fiable en production. Le mode strict vous permet de corriger les bugs avant que vos utilisateurs ne les signalent.

<Note>

Le mode strict active les vérifications suivantes en mode de développement :

> - Vos composants seront [rendus une fois de plus](#fixing-bugs-found-by-double-rendering-in-development) afin de trouver les bugs causés par des rendus impurs.

Reviewer: ```suggestion
- Vos composants feront [un rendu supplémentaire](#fixing-bugs-found-by-double-rendering-in-development) afin de trouver les bugs causés par des rendus impurs.
```

> - Vos composants [réexécuteront les Effets une fois de plus](#fixing-bugs-found-by-re-running-effects-in-development) afin de détecter les bugs causés par l'absence de nettoyage d'Effet.

Reviewer: ```suggestion
- Vos composants [exécuteront les Effets une fois supplémentaire](#fixing-bugs-found-by-re-running-effects-in-development) afin de détecter les bugs causés par l'absence de nettoyage d'Effet.
```

}
```

Dans cet exemple, les vérifications du mode strict ne s'exécuteront pas sur les composants `Header` et `Footer`. Cependant, elles s'exécuteront sur `Sidebar` et `Content`, ainsi que sur tous les composants qu'ils contiennent, peu importe la profondeur à laquelle ils se trouvent.

---

### Corriger les bugs trouvés par le double rendu en développement {/*fixing-bugs-found-by-double-rendering-in-development*/}

> [React part du principe que chaque composant que vous écrivez est une fonction pure.](/learn/keeping-components-pure) Ça signifie que vos composants React doivent toujours renvoyer le même JSX pour les mêmes entrées (props, état et contexte).

Reviewer: ```suggestion
[React part du principe que chaque composant que vous écrivez est une fonction pure](/learn/keeping-components-pure). Ça signifie que vos composants React doivent toujours renvoyer le même JSX pour les mêmes entrées (props, état et contexte).
```

Les composants qui ne respectent pas cette règle peuvent se comporter de façon imprévisible et occasionner des bugs. Pour vous aider à trouver du code accidentellement impur, le mode strict appelle certaines de vos fonctions (seulement celles qui doivent être pures) **deux fois en développement**. Ça inclut :

> - Le corps de la fonction de votre composant (seulement la logique de haut niveau, ce qui exclut le code contenu dans les gestionnaires d'événements).

Reviewer: ```suggestion
- Le corps de votre fonction composant (seulement la logique du niveau racine, ce qui exclut le code contenu dans les gestionnaires d'événements).
```

> - Les fonctions que vous donnez à [`useState`](/reference/react/useState), [les fonctions `set`](/reference/react/useState#setstate), [`useMemo`](/reference/react/useMemo) ou [`useReducer`](/reference/react/useReducer).

Reviewer: ```suggestion
- Les fonctions que vous passez à [`useState`](/reference/react/useState), aux [fonctions `set`](/reference/react/useState#setstate), à [`useMemo`](/reference/react/useMemo) ou à [`useReducer`](/reference/react/useReducer).
```

> - Certaines méthodes des composants écrits avec des classes comme [`constructor`](/reference/react/Component#constructor), [`render`](/reference/react/Component#render), [`shouldComponentUpdate`](/reference/react/Component#shouldcomponentupdate) ([voir la liste exhaustive](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects)).

Reviewer: ```suggestion
- Certaines méthodes des composants à bases de classes comme [`constructor`](/reference/react/Component#constructor), [`render`](/reference/react/Component#render), [`shouldComponentUpdate`](/reference/react/Component#shouldcomponentupdate) ([voir la liste complète](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects)).
```

> Si une fonction est pure, l'exécuter deux fois ne change pas son comportement, car une telle fonction produit le même résultat à chaque fois. Cependant, si une fonction est impure (elle modifie par exemple la donnée qu'elle reçoit), l'exécuter deux fois devrait se remarquer (c'est ce qui la rend impure !). Ça vous aide à déceler et corriger les bugs plus rapidement.

Reviewer: ```suggestion
Si une fonction est pure, l'exécuter deux fois ne change pas son comportement, car une telle fonction produit le même résultat à chaque fois. Cependant, si une fonction est impure (elle modifie par exemple la donnée qu'elle reçoit), l'exécuter deux fois devrait se remarquer (c'est ce qui la rend impure !). Ça vous aide à détecter et corriger les bugs plus rapidement.
```

import StoryTray from './StoryTray.js';

let initialStories = [
  { id: 0, label: "L’histoire d’Ankit" },

>   { id: 1, label: "L’histoire de Taylor" },

Reviewer: ```suggestion
  { id: 1, label: "L’histoire de Clara" },
```

cf. keeping-components-pure.md, qui a déjà tout ça.

Author: J'ai hésité pour Taylor, car il me semblait qu'on avait rien pour Ankit, donc j'ai laissé tout tel quel...

</Sandpack>

> Il y a une erreur dans le code ci-dessus. Il est cependant facile de passer à côté car l'affichage initial semble correct.

Reviewer: Très bien, "passer à côté" 👏

Author: Merci, j'ai réécrit 3 fois la phrase pour trouver la bonne tournure 😉

> Cette erreur devient bien plus notable si le composant `StoryTray` est rendu plusieurs fois. Par exemple, faisons un nouveau rendu de `StoryTray` avec une couleur de fond différente à chaque fois que vous le survolez :

Reviewer: ```suggestion
Cette erreur devient bien plus notable si le composant `StoryTray` fait son rendu plusieurs fois. Par exemple, faisons un nouveau rendu de `StoryTray` avec une couleur de fond différente à chaque fois que vous le survolez :
```

import StoryTray from './StoryTray.js';

let initialStories = [
  { id: 0, label: "L’histoire d’Ankit" },

>   { id: 1, label: "L’histoire de Taylor" },

Reviewer: ```suggestion
  { id: 1, label: "L’histoire de Clara" },
```

</Sandpack>

> Remarquez qu'à chaque fois que vous survolez le composant `StoryTray`, « Créer une histoire » est ajouté à la liste. L'intention du code n'était de l'ajouter qu'une seule fois à la fin. Cependant, `StoryTray` modifie directement le tableau `stories` des props. Chaque fois que `StoryTray` fait son rendu, il ajoute « Créer une histoire » à la fin de ce tableau. En d'autres termes, `StoryTray` n'est pas une fonction pure — l'exécuter plusieurs fois produit des résultats différents.

Reviewer: ```suggestion
Remarquez qu'à chaque fois que vous survolez le composant `StoryTray`, ça ajoute « Créer une histoire » à la liste. Le code ne visait qu'à l'ajouter une seule fois à la fin. Cependant, `StoryTray` modifie directement le tableau `stories` des props. Chaque fois que `StoryTray` fait son rendu, il ajoute « Créer une histoire » à la fin de ce tableau. En d'autres termes, `StoryTray` n'est pas une fonction pure — l'exécuter plusieurs fois produit des résultats différents.
```

Pour corriger ce problème, vous pouvez faire une copie du tableau et modifier cette copie plutôt que l'original :

```js {2}
export default function StoryTray({ stories }) {

>   const items = stories.slice(); // Cloner le tableau
>   // ✅ Bien : ajouter dans un nouveau tableau

Reviewer: ```suggestion
  const items = stories.slice(); // Copier le tableau
  // ✅ Correct : ajouter dans un nouveau tableau
```

- Même commentaire que dans l'autre page de cet exemple
- J'évite "Bon" / "Mauvais" au profit de "Correct" et "Erroné", non connotés.

  items.push({ id: 'create', label: 'Créer une histoire' });
```

> Ça rendrait [la fonction `StoryTray` pure](/learn/keeping-components-pure). À chaque appel, elle ne modifierait que la copie du tableau et n'affecterait aucun objet ou variable externe. Ça résoud le bug, mais vous avez dû faire en sorte que le composant soit rendu plus souvent avant qu'il ne devienne évident que son comportement n'est pas correct.

Reviewer: ```suggestion
Ça rendrait [la fonction `StoryTray` pure](/learn/keeping-components-pure). À chaque appel, elle ne modifierait que la copie du tableau et n'affecterait aucun objet ou variable externe. Ça résout le bug, mais vous avez dû faire en sorte que le composant fasse plus souvent son rendu avant qu'il ne devienne évident que son comportement n'était pas correct.
```

import StoryTray from './StoryTray.js';

let initialStories = [
  { id: 0, label: "L’histoire d’Ankit" },

>   { id: 1, label: "L’histoire de Taylor" },

Reviewer: ```suggestion
  { id: 1, label: "L’histoire de Clara" },
```

</Sandpack>

> **Le mode strict appelle *toujours* votre fonction de rendu deux fois, afin que vous puissiez voir le problème directement** (« Créer une histoire » apparaît deux fois). Ça vous permet de détecter ce genre d'erreur plus tôt dans le processus. Lorsque vous corrigez votre composant pour qu'il soit rendu en mode strict, vous corrigez *également* de nombreux bugs potentiels en production, telle que la fonctionnalité de survol précédente :

Reviewer: ```suggestion
**Le mode strict appelle *toujours* votre fonction de rendu deux fois, afin que vous puissiez voir le problème immédiatement** (« Créer une histoire » apparaît deux fois). Ça vous permet de détecter ce genre d'erreur plus tôt dans le processus. Lorsque vous corrigez votre composant pour qu'il fasse des rendus corrects en mode strict, vous corrigez *également* de nombreux bugs potentiels en production, telle que la fonctionnalité de survol précédente :
```

import StoryTray from './StoryTray.js';

let initialStories = [
  { id: 0, label: "L’histoire d’Ankit" },

>   { id: 1, label: "L’histoire de Taylor" },

Reviewer: ```suggestion
  { id: 1, label: "L’histoire de Clara" },
```

export default function StoryTray({ stories }) {
  const [isHover, setIsHover] = useState(false);

>   const items = stories.slice(); // Cloner le tableau

Reviewer: ```suggestion
  const items = stories.slice(); // Copier le tableau
```

</Sandpack>

Sans le mode strict, il était facile de passer à côté du bug jusqu'à ce que vous ajoutiez d'autres rendus. Le mode strict fait apparaître le même bug immédiatement. Ce mode vous aide à trouver les bugs avant que vous ne les poussiez à votre équipe et à vos utilisateurs.

> [Apprenez-en davantage sur la façon de garder les composants purs.](/learn/keeping-components-pure)

Reviewer: ```suggestion
[Apprenez-en davantage sur la façon de garder les composants purs](/learn/keeping-components-pure).
```

<Note>

> Si vous avez installé [React DevTools](/learn/react-developer-tools), tous les appels à `console.log` lors du second appel apparaîtront légèrement estompés. React DevTools propose également un paramètre (désactivé par défaut) pour les supprimer complétement.

Reviewer: ```suggestion
Si vous avez installé [les outils de développement React](/learn/react-developer-tools), tous les appels à `console.log` lors du second appel apparaîtront légèrement estompés. Ces outils proposent également un paramètre (désactivé par défaut) pour les masquer complètement.
```

</Note>

---

### Corriger les bugs trouvés en réexécutant les Effets en développement {/*fixing-bugs-found-by-re-running-effects-in-development*/}

Le mode strict est également utile pour trouver des bugs dans les [Effets](/learn/synchronizing-with-effects).

> Chaque Effet a du code d'initialisation et peut avoir du code de nettoyage. Normalement, React appelle le code d'initialisation quand le composant *est monté* (quand il est ajouté à l'écran), et appelle le code de nettoyage quand celui-ci est *démonté* (il est enlevé de l'écran). React appelle ensuite le nettoyage et l'initialisation à nouveau si l'une des dépendances de l'Effet a changé depuis le dernier rendu.

Reviewer: ```suggestion
Chaque Effet a du code d'initialisation et peut avoir du code de nettoyage. Normalement, React appelle le code d'initialisation quand le composant *est monté* (quand il est ajouté à l'écran), et appelle le code de nettoyage quand le composant est *démonté* (il est enlevé de l'écran). React appelle ensuite le nettoyage et l'initialisation à nouveau si l'une des dépendances de l'Effet a changé depuis le dernier rendu.

Quand le mode strict est activé, React exécutera **un cycle d'initialisation + nettoyage supplémentaire en développement pour chaque Effet**.  Ça peut surprendre, mais ça aide à détecter des bugs subtils qu'il est difficile de repérer manuellement.
```

Et donc tu as viré un paragraphe entier en fait ? 😨

Author: ça fait plusieurs fois que ça m'arrive. Je ne sais pas ce que je fais dans mon VS Code, parce que _promis_ je ne le fais pas exprès !

Et sachant ça, j'avais pourtant fait une passe (trop) rapide dans mon éditeur pour voir si j'avais supprimé des paragraphes. Merci pour ta vigilance, je revérifierai mieux la prochaine fois (surtout qu'on devrait avoir le même nombre de lignes entre la VF et la VO !!)

</Sandpack>

> Il y a un problème avec ce code, mais ce n'est pas forcément évident.

Reviewer: ```suggestion
Il y a un problème avec ce code, mais ça ne saute pas aux yeux.
```

> Pour rendre le problème plus évident, ajoutons une fonctionnalité. Dans l'exemple ci-dessous, `roomId` n'est pas codée en dur. L'utilisateur peut en effet choisir dans une liste déroulante le `roomId` auquel il souhaite se connecter. Appuyez sur « Ouvrir le salon » puis sélectionnez une à une les différents salons de discussion. Surveillez le nombre de connexions actives dans la console :

Reviewer: ```suggestion
Pour rendre le problème plus évident, ajoutons une fonctionnalité. Dans l'exemple ci-dessous, `roomId` n'est pas codé en dur. L'utilisateur peut en effet choisir dans une liste déroulante le `roomId` auquel il souhaite se connecter. Appuyez sur « Ouvrir le salon » puis sélectionnez un à un les différents salons de discussion. Surveillez le nombre de connexions actives dans la console :
```

</Sandpack>

> Vous remarquerez que le nombre de connexions ouvertes n'arrête pas de grandir. Dans une vraie appli, ça causerait des problèmes de performances et de réseau. Le problème vient [de l'absence de fonction de nettoyage dans votre Effet](/learn/synchronizing-with-effects#step-3-add-cleanup-if-needed) :

Reviewer: ```suggestion
Vous remarquerez que le nombre de connexions ouvertes ne cesse de grandir. Dans une véritable appli, ça causerait des problèmes de performances et de réseau. Le problème vient [de l'absence de fonction de nettoyage dans votre Effet](/learn/synchronizing-with-effects#step-3-add-cleanup-if-needed) :
```

</Sandpack>

Remarquez que le nombre de connexions actives dans la console cesse de grandir maintenant.

Sans le mode strict, il était facile de passer à côté du fait que l'Effet nécessitait une fonction de nettoyage. En exécutant *initialisation → nettoyage → initialisation* plutôt que *initialisation* de votre Effet en développement, le mode strict a rendu l'absence de fonction de nettoyage plus visible.

> [Apprenez-en davantage sur l'implémentation de fonction de nettoyage des Effets.](/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development)

Reviewer: ```suggestion
[Apprenez-en davantage sur l'implémentation de fonction de nettoyage des Effets](/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development).
```

---

> ### Corriger les alertes de dépréciation acivées par le mode strict {/*fixing-deprecation-warnings-enabled-by-strict-mode*/}

Reviewer: ```suggestion
### Corriger les alertes de dépréciation activées par le mode strict {/*fixing-deprecation-warnings-enabled-by-strict-mode*/}
```

> React alerte si certains composants quelque part à l'intérieur de l'arbre de `<StrictMode>` utilise l'une de ces APIs dépréciées :

Reviewer: ```suggestion
React alerte si certains composants quelque part à l'intérieur de l'arbre de `<StrictMode>` utilisent l'une de ces API dépréciées :
```

> * [`findDOMNode`](/reference/react-dom/findDOMNode). [Voir les alternatives.](https://reactjs.org/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage)

Reviewer: ```suggestion
* [`findDOMNode`](/reference/react-dom/findDOMNode). [Voir les alternatives](/reference/react-dom/findDOMNode#alternatives).
```

Super la VO et son vieux lien pourri…

> * Les méthodes de classe du cycle de vie `UNSAFE_` telle que [`UNSAFE_componentWillMount`](/reference/react/Component#unsafe_componentwillmount). [Voir les alternatives.](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#migrating-from-legacy-lifecycles) 

Reviewer: ```suggestion
* Les méthodes de cycle de vie `UNSAFE_` des classes telle que [`UNSAFE_componentWillMount`](/reference/react/Component#unsafe_componentwillmount). [Voir les alternatives](https://legacy.reactjs.org/blog/2018/03/27/update-on-async-rendering.html#migrating-from-legacy-lifecycles).
```

> * Les anciens contextes ([`childContextTypes`](/reference/react/Component#static-childcontexttypes), [`contextTypes`](/reference/react/Component#static-contexttypes) et [`getChildContext`](/reference/react/Component#getchildcontext)). [Voir les alternatives.](/reference/react/createContext)

Reviewer: ```suggestion
* Les anciens contextes ([`childContextTypes`](/reference/react/Component#static-childcontexttypes), [`contextTypes`](/reference/react/Component#static-contexttypes) et [`getChildContext`](/reference/react/Component#getchildcontext)). [Voir les alternatives](/reference/react/createContext).
```

> * Les anciennes refs ([`this.refs`](/reference/react/Component#refs)). [Voir les alternatives.](https://reactjs.org/docs/strict-mode.html#warning-about-legacy-string-ref-api-usage)

Reviewer: ```suggestion
* Les anciennes refs textuelles ([`this.refs`](/reference/react/Component#refs)). [Voir les alternatives](https://legacy.reactjs.org/docs/strict-mode.html#warning-about-legacy-string-ref-api-usage).
```

> Ces APIs sont avant tout utilisées dans les anciens [composants de classe](/reference/react/Component) et n'apparaîssent que rarement dans les applis modernes.

Reviewer: ```suggestion
Ces API sont avant tout utilisées dans les anciens [composants à base de classes](/reference/react/Component) et n'apparaissent que rarement dans les applis modernes.
```
